module/zfs/vdev.c

```diff
@@ -3670,6 +3671,116 @@ vdev_deadman(vdev_t *vd)
 	}
 }
 
+/*
+ * Implements the per-vdev portion of manual TRIM. The function passes over
+ * all metaslabs on this vdev and performs a metaslab_trim_all on them. It's
+ * also responsible for rate-control if spa_man_trim_rate is non-zero.
+ *
+ * If fulltrim is set, metaslabs without spacemaps are also trimmed.
```

Comment:
I guess that's what `-f` does.  So even without `-f`, it trims space that was never written but is part of metaslabs that have had some allocations.  IMO, it isn't really worth making this distinction.  I'd say just always trim all unallocated space (i.e. always do the `-f` behavior).